cli/daemon/engine/trace2/sqlite/write.go

```diff
@@ -29,10 +35,103 @@ type Store struct {
 
 var _ trace2.Store = (*Store)(nil)
 
+func scanRows[T any](rows *sql.Rows) ([]T, error) {
+	defer rows.Close()
+	var out []T
+	for rows.Next() {
+		var v T
+		err := rows.Scan(&v)
+		if err != nil {
+			return nil, err
+		}
+		out = append(out, v)
+	}
+	return out, nil
+}
+
+func (s *Store) startCleaner(ctx context.Context, freq time.Duration, triggerAt, eventsToKeep, batchSize int) {
+	go func() {
+		for {
+			timer := time.NewTimer(freq)
+			select {
+			case <-ctx.Done():
+				return
+			case <-timer.C:
+				log.Info().Msg("initiating trace event cleanup sweep")
+				rows, err := s.db.QueryContext(ctx, "SELECT app_id FROM trace_event GROUP BY app_id HAVING COUNT(distinct trace_id) > ?", triggerAt)
+				if err != nil {
+					log.Error().Err(err).Msg("failed to get app ids")
+					continue
+				}
+				appIDs, err := scanRows[string](rows)
+				if err != nil {
+					log.Error().Err(err).Msg("failed to scan app ids")
+					continue
+				}
+				for _, appID := range appIDs {
+					row := s.db.QueryRowContext(ctx, `
+						WITH latest_events AS (
+							SELECT trace_id, min(id) as id FROM trace_event WHERE app_id = ? GROUP BY 1 ORDER BY 2 DESC LIMIT ?
+						) SELECT min(id) FROM latest_events;
+					`, appID, eventsToKeep)
+					var traceID int64
+					err := row.Scan(&traceID)
+					if err != nil {
+						log.Error().Err(err).Msg("failed to get trace id")
+						continue
+					}
+					rows, err := s.db.QueryContext(ctx, "SELECT DISTINCT trace_id FROM trace_event WHERE app_id = ? AND id < ? ORDER BY id DESC LIMIT ?", appID, traceID, batchSize)
+					if err != nil {
+						log.Error().Err(err).Msg("failed to get old trace ids")
+						continue
+					}
+					traceIDs, err := scanRows[string](rows)
+					if len(traceIDs) == 0 {
+						continue
+					}
+					idArgs := strings.Join(fns.Map(traceIDs, pq.QuoteLiteral), ",")
+					res, err := s.db.ExecContext(ctx, "DELETE FROM trace_event WHERE app_id = ? AND trace_id IN ("+idArgs+")", appID)
```

Comment:
In Postgres at least you can do `trace_id = ANY(?)` and pass in a string slice, instead of having to construct the comma-separated string arg by hand. Not sure if that works in sqlite though

Reply:
yeah, it doesn't unfortunately :(. There's a bunch of libs solving this in different ways, but I figured this was good enough.